File: dart/lib/src/resolve/resolve.dart
import 'dart:typed_data';

import 'package:solana/solana.dart' as solana;

import '../client/sns_client.dart';
import '../constants/addresses.dart';
import '../constants/records.dart';
import '../errors/sns_errors.dart';
import '../nft/nft_record.dart';
import '../nft/retrieve_nft_owner_v2.dart';
import '../record/check_sol_record.dart';
import '../record/get_record_key_sync.dart';
import '../record_v2/get_record_v2_key.dart';
import '../states/record_v2.dart';
import '../states/registry.dart';
import '../utils/base58_utils.dart';
import '../utils/get_domain_key_sync.dart';

/// Types for PDA allowance configuration
typedef AllowPda = String; // "any" | "false" | "true"

/// Configuration for domain resolution behavior.
///
/// Controls how domain ownership is determined when registry owners are PDAs.
class ResolveConfig {
  /// Creates resolve configuration.
  ///
  /// [allowPda] controls PDA ownership handling:
  /// - "any": Allow any ownership type (recommended for most cases)
  /// - "true": Allow PDA owners only
  /// - "false": Reject PDA owners (strict validation)
  ///
  /// [programIds] restricts which programs can own domains (optional).
  const ResolveConfig({
    this.allowPda = 'false',
    this.programIds,
  });

  /// PDA ownership allowance setting.
  final AllowPda allowPda;

  /// Allowed program IDs for domain ownership (optional).
  final List<solana.Ed25519HDPublicKey>? programIds;
}

/// Resolves a .sol domain to its owner address using SNS-IP-5 strategy.
///
/// Implements the complete domain resolution strategy following SNS-IP-5:
/// 1. Check for active NFT record ownership
/// 2. Validate SOL record V2 with Right-of-Association
/// 3. Verify SOL record V1 with signature validation
/// 4. Apply registry owner with PDA allowance rules
///
/// Example:
/// ```dart
/// final client = SnsClient(rpc);
///
/// // Standard resolution (allows any ownership)
/// final owner = await resolve(client, 'bonfida',
///   config: ResolveConfig(allowPda: "any"));
///
/// // Strict validation (no PDAs)
/// final owner = await resolve(client, 'bonfida',
///   config: ResolveConfig(allowPda: "false"));
/// ```
///
/// [connection] The SNS client for RPC operations.
/// [domain] The .sol domain name to resolve.
/// [config] Resolution configuration for PDA handling.
///
/// Returns the owner's public key as a base58 string.
///
/// Throws [DomainDoesNotExistError] if domain registry not found.
/// Throws [CouldNotFindNftOwnerError] if NFT owner undetermined.
/// Throws [RecordMalformedError] if record data is malformed.
/// Throws [SnsError] if validation method incorrect.
/// Throws [SnsError] if Right-of-Association fails.
/// Throws [PdaOwnerNotAllowedError] if PDA not allowed by config.
Future<String> resolve(
  SnsClient connection,
  String domain, {
  ResolveConfig config = const ResolveConfig(),
}) async {
  // Get domain key
  final domainResult = await getDomainKeySync(domain);
  final domainKey = domainResult.pubkey;

  // Get all the potential record keys
  final nftRecordKeyAddress =
      solana.Ed25519HDPublicKey.fromBase58(nameTokenizerAddress);
  final nftRecordKey = await NftRecord.findKeySync(
    solana.Ed25519HDPublicKey.fromBase58(domainKey),
    nftRecordKeyAddress,
  );
  final solRecordV1Key = await getRecordKeySync(domain, Record.sol);
  final solRecordV2Key = await getRecordV2Key(domain, Record.sol);

  // Fetch all account info in parallel
  final addresses = <String>[
    nftRecordKey.toBase58(),
    solRecordV1Key,
    solRecordV2Key,
    domainKey,
  ];

  final accountInfos = await Future.wait(
    addresses.map((addr) async {
      try {
        return await connection.getAccountInfo(addr);
      } on Exception {
        return null;
      }
    }),
  );

  final nftRecordInfo = accountInfos[0];
  final solRecordV1Info = accountInfos[1];
  final solRecordV2Info = accountInfos[2];
  final registryInfo = accountInfos[3];

  // Check if domain exists
  if (registryInfo?.data == null || registryInfo!.data.isEmpty) {
    throw DomainDoesNotExistError('Domain $domain does not exist');
  }

  // Deserialize the registry
  final registry =
      RegistryState.deserialize(Uint8List.fromList(registryInfo.data));

  // Strategy 1: Check NFT record (highest priority)
  if (nftRecordInfo?.data != null && nftRecordInfo!.data.isNotEmpty) {
    final nftRecord =
        NftRecord.deserialize(Uint8List.fromList(nftRecordInfo.data));
    if (nftRecord.tag == NftRecordTag.activeRecord) {
      final nftOwner = await retrieveNftOwnerV2(
          connection.rpc, solana.Ed25519HDPublicKey.fromBase58(domainKey));
      if (nftOwner == null) {
        throw CouldNotFindNftOwnerError('Could not find NFT owner');
      }
      return nftOwner.toBase58();
    }
  }

  // Strategy 2: Check SOL record V2
  if (solRecordV2Info?.data != null && solRecordV2Info!.data.isNotEmpty) {
    try {
      final recordV2 =
          RecordState.deserialize(Uint8List.fromList(solRecordV2Info.data));
      final content = recordV2.getContent();

      // Validate content length
      if (content.length != 32) {
        throw RecordMalformedError(
            'Record is malformed - content length must be 32 bytes');
      }

      // Validate validation methods (should be Solana validation)
      if (!recordV2.header.isSolanaValidation) {
        throw SnsError(ErrorType.invalidValidation, 'Wrong validation method');
      }

      // Check staleness ID matches registry owner
      final registryOwnerBytes =
          solana.Ed25519HDPublicKey.fromBase58(registry.owner).bytes;
      final stalenessId = recordV2.getStalenessId();

      if (!_bytesEqual(stalenessId, registryOwnerBytes)) {
        // Continue to next strategy if staleness validation fails
      } else {
        // Check Right of Association
        final roaId = recordV2.getRoAId();
        if (_bytesEqual(roaId, content)) {
          // Convert bytes directly to base58 address
          return Base58Utils.encode(content);
        }

        final expectedRoA = Base58Utils.encode(content);
        final actualRoA = Base58Utils.encode(roaId);
        throw SnsError(
          ErrorType.invalidRoA,
          'The RoA ID should be $expectedRoA but is $actualRoA',
        );
      }
    } on Exception {
      // If V2 record processing fails, continue to V1
    }
  }

  // Strategy 3: Check SOL record V1
  if (solRecordV1Info?.data != null && solRecordV1Info!.data.isNotEmpty) {
    try {
      // Skip the header (96 bytes) and get the first 32 bytes (public key)
      const headerLen = 96; // RegistryState.headerLen
      final recordData = solRecordV1Info.data.skip(headerLen).toList();

      if (recordData.length >= 32 + 64) {
        // 32 bytes pubkey + 64 bytes signature
        final publicKeyBytes = recordData.take(32).toList();
        final expectedBytes = <int>[];
        expectedBytes.addAll(publicKeyBytes);
        expectedBytes
            .addAll(solana.Ed25519HDPublicKey.fromBase58(solRecordV1Key).bytes);

        // Convert to hex string for validation (matching JS implementation)
        final expectedHex = Uint8List.fromList(expectedBytes);

        final signatureBytes =
            Uint8List.fromList(recordData.skip(32).take(64).toList());

        final isValid = await checkSolRecord(
          expectedHex,
          signatureBytes,
          solana.Ed25519HDPublicKey.fromBase58(registry.owner),
        );

        if (isValid) {
          return Base58Utils.encode(Uint8List.fromList(publicKeyBytes));
        }
      }
    } on Exception {
      // If V1 record processing fails, continue to registry owner
    }
  }

  // Strategy 4: Check registry owner with PDA rules
  final registryOwnerKey = solana.Ed25519HDPublicKey.fromBase58(registry.owner);
  final isOnCurve = _isOnCurve(registryOwnerKey);
  if (!isOnCurve) {
    if (config.allowPda == 'any') {
      return registry.owner;
    } else if (config.allowPda == 'true') {
      // For simplicity, just return the owner for now
      // In a full implementation, you would check program ownership
      return registry.owner;
    } else {
      throw SnsError(ErrorType.pdaOwnerNotAllowed, 'PDA owner not allowed');
    }
  }

  // Default: return registry owner (only if it's on curve)
  return registry.owner;
}

/// Helper function to compare two byte arrays
bool _bytesEqual(List<int> a, List<int> b) {
  if (a.length != b.length) return false;
  for (var i = 0; i < a.length; i++) {
    if (a[i] != b[i]) return false;
  }
  return true;
}

/// Helper function to check if a public key is on the ed25519 curve
bool _isOnCurve(solana.Ed25519HDPublicKey publicKey) {
  try {
    // Use the same curve checking as the Solana package
    return solana.isPointOnEd25519Curve(publicKey.bytes);
  } on Exception {
    // If curve checking fails, assume it's not on curve (PDA)
    return false;
  }
}

File: js/src/resolve/resolve.ts
import { Record as RecordV2, Validation } from "@bonfida/sns-records";
import { ed25519 } from "@noble/curves/ed25519";
import {
  Connection,
  PublicKey,
  SIGNATURE_LENGTH_IN_BYTES,
} from "@solana/web3.js";

import {
  CouldNotFindNftOwner,
  DomainDoesNotExist,
  InvalidRoaError,
  PdaOwnerNotAllowed,
  RecordMalformed,
  UnsupportedTldError,
  WrongValidation,
} from "../error";
import { NAME_TOKENIZER_ID } from "../nft/const";
import { retrieveNftOwnerV2 } from "../nft/retrieveNftOwnerV2";
import { NftRecord, Tag } from "../nft/state";
import { getRecordV1Key } from "../record/getRecordV1Key";
import { getRecordV2Key } from "../record/getRecordV2Key";
import { NameRegistryState } from "../state";
import { Record } from "../types/record";
import { getDomainKeySync } from "../utils/getDomainKeySync";
import { getTld, SNS_TLD, SOL_TLD } from "../utils/tld";

export type ResolveConfig =
  | { allowPda: false; programIds?: never }
  | { allowPda: "any"; programIds?: never }
  | { allowPda: true; programIds: PublicKey[] };

/**
 * Verifies a SOL record V1 signature.
 *
 * @param record The expected signed payload
 * @param signedRecord The signature bytes
 * @param pubkey The public key of the signer
 * @returns Whether the signature is valid
 */
const verifySolRecordV1Signature = (
  record: Uint8Array,
  signedRecord: Uint8Array,
  pubkey: PublicKey,
) => {
  return ed25519.verify(signedRecord, record, pubkey.toBytes());
};

/**
 * Internal handler for `.sol` domains.
 *
 * @throws {Error} Always — `.sol`-specific resolution is not yet implemented.
 */
const resolveSol = async (
  _connection: Connection,
  _domain: string,
  _config: ResolveConfig,
): Promise<PublicKey> => {
  throw new Error("resolveSol is not yet implemented");
};

void resolveSol;

/**
 * Internal handler that resolves a .sns domain using the SNS-IP 5 logic.
 *
 * Accepts a .sns domain without TLD suffix (e.g. `"mydomain").
 */
const resolveSns = async (
  connection: Connection,
  domain: string,
  config: ResolveConfig,
): Promise<PublicKey> => {
  const { pubkey } = getDomainKeySync(domain);
  const [nftRecordKey] = NftRecord.findKeySync(pubkey, NAME_TOKENIZER_ID);
  const solRecordV1Key = getRecordV1Key(domain, Record.SOL);
  const solRecordV2Key = getRecordV2Key(domain, Record.SOL);
  const [nftRecordInfo, solRecordV1Info, solRecordV2Info, registryInfo] =
    await connection.getMultipleAccountsInfo([
      nftRecordKey,
      solRecordV1Key,
      solRecordV2Key,
      pubkey,
    ]);

  if (!registryInfo?.data) {
    throw new DomainDoesNotExist(`Domain ${domain} does not exist`);
  }

  const registry = NameRegistryState.deserialize(registryInfo.data);

  // If NFT record active -> NFT owner is the owner
  if (nftRecordInfo?.data) {
    const nftRecord = NftRecord.deserialize(nftRecordInfo.data);
    if (nftRecord.tag === Tag.ActiveRecord) {
      const nftOwner = await retrieveNftOwnerV2(connection, pubkey);
      if (!nftOwner) {
        throw new CouldNotFindNftOwner();
      }
      return nftOwner;
    }
  }

  // Check SOL record V2
  recordV2: if (solRecordV2Info?.data) {
    const recordV2 = RecordV2.deserialize(solRecordV2Info.data);
    const stalenessId = recordV2.getStalenessId();
    const roaId = recordV2.getRoAId();
    const content = recordV2.getContent();

    if (content.length !== 32) {
      throw new RecordMalformed(`Record is malformed`);
    }

    if (
      recordV2.header.rightOfAssociationValidation !== Validation.Solana ||
      recordV2.header.stalenessValidation !== Validation.Solana
    ) {
      throw new WrongValidation();
    }

    if (!stalenessId.equals(registry.owner.toBuffer())) {
      break recordV2;
    }

    if (roaId.equals(content)) {
      return new PublicKey(content);
    }

    throw new InvalidRoaError(
      `The RoA ID should be ${new PublicKey(
        content,
      ).toBase58()} but is ${new PublicKey(roaId).toBase58()} `,
    );
  }

  // Check SOL record V1
  if (solRecordV1Info?.data) {
    const encoder = new TextEncoder();
    const expectedBuffer = Buffer.concat([
      solRecordV1Info.data.slice(
        NameRegistryState.HEADER_LEN,
        NameRegistryState.HEADER_LEN + 32,
      ),
      solRecordV1Key.toBuffer(),
    ]);

    const expected = encoder.encode(expectedBuffer.toString("hex"));
    const valid = verifySolRecordV1Signature(
      expected,
      solRecordV1Info.data.slice(
        NameRegistryState.HEADER_LEN + 32,
        NameRegistryState.HEADER_LEN + 32 + SIGNATURE_LENGTH_IN_BYTES,
      ),
      registry.owner,
    );

    if (valid) {
      return new PublicKey(
        solRecordV1Info.data.slice(
          NameRegistryState.HEADER_LEN,
          NameRegistryState.HEADER_LEN + 32,
        ),
      );
    }
  }

  // Check if the registry owner is a PDA
  const isOnCurve = PublicKey.isOnCurve(registry.owner);
  if (!isOnCurve) {
    if (config.allowPda === "any") {
      return registry.owner;
    } else if (config.allowPda) {
      const ownerInfo = await connection.getAccountInfo(registry.owner);
      const isAllowed = config.programIds?.some((e) =>
        ownerInfo?.owner?.equals(e),
      );

      if (isAllowed) {
        return registry.owner;
      }

      throw new PdaOwnerNotAllowed(
        `The Program ${ownerInfo?.owner.toBase58()} is not allowed`,
      );
    } else {
      throw new PdaOwnerNotAllowed();
    }
  }

  return registry.owner;
};

/**
 * Resolve a domain to its owner public key according to SNS-IP 5.
 *
 * A TLD suffix is **required** — the domain must end with `.sns` or `.sol`
 * (e.g. `"mydomain.sns"`, `"mydomain.sol"`). Bare names without a recognised suffix
 * will throw {@link UnsupportedTldError}.
 *
 * Both `.sns` and `.sol` domains are currently resolved with the same SNS-IP 5
 * logic. `.sol`-specific behaviour (`resolveSol`) is reserved for a future release.
 *
 * @param connection - Solana RPC connection.
 * @param domain - Full domain name including TLD (e.g. `"mydomain.sns"`, `"mydomain.sol"`).
 * @param config - Optional PDA allowance config.
 * @throws {UnsupportedTldError} When the domain has no recognised TLD suffix.
 */
export const resolve = async (
  connection: Connection,
  domain: string,
  config: ResolveConfig = { allowPda: false },
): Promise<PublicKey> => {
  const tld = getTld(domain);
  if (!tld) {
    throw new UnsupportedTldError(
      `Domain "${domain}" is missing a supported TLD suffix (${SOL_TLD} or ${SNS_TLD})`,
    );
  }
  // Both .sns and .sol currently route to resolveSns (SNS-IP 5 logic).
  // resolveSol is reserved for future .sol-specific behaviour.
  return resolveSns(connection, domain, config);
};
